lib/lfrfid/protocols/protocol_paradox.c

```diff
@@ -0,0 +1,211 @@
+    }
+
+    return result;
+};
+
+static void protocol_paradox_encode(const uint8_t* decoded_data, uint8_t* encoded_data) {
+    memset(encoded_data, 0, PARADOX_ENCODED_DATA_SIZE);
+
+    // preamble
+    bit_lib_set_bits(encoded_data, 0, 0b00001111, 8);
+
+    for(size_t i = 0; i < 44; i++) {
+        if(bit_lib_get_bit(decoded_data, i)) {
+            bit_lib_set_bits(encoded_data, PARADOX_PREAMBLE_LENGTH + i * 2, 0b10, 2);
+        } else {
+            bit_lib_set_bits(encoded_data, PARADOX_PREAMBLE_LENGTH + i * 2, 0b01, 2);
+        }
+    }
+};
+
+bool protocol_paradox_encoder_start(ProtocolParadox* protocol) {
+    protocol_paradox_encode(protocol->data, (uint8_t*)protocol->encoded_data);
+    protocol->encoder.encoded_index = 0;
+    fsk_osc_reset(protocol->encoder.fsk_osc);
+    return true;
+};
+
+LevelDuration protocol_paradox_encoder_yield(ProtocolParadox* protocol) {
+    bool level;
+    uint32_t duration;
+
+    bool bit = bit_lib_get_bit(protocol->encoded_data, protocol->encoder.encoded_index);
+    bool advance = fsk_osc_next_half(protocol->encoder.fsk_osc, bit, &level, &duration);
+
+    if(advance) {
+        bit_lib_increment_index(protocol->encoder.encoded_index, PARADOX_ENCODED_BIT_SIZE);
+    }
+    return level_duration_make(level, duration);
+};
+
+void protocol_paradox_render_data(ProtocolParadox* protocol, string_t result) {
+    uint8_t* decoded_data = protocol->data;
+    uint8_t fc = bit_lib_get_bits(decoded_data, 10, 8);
+    uint16_t card_id = bit_lib_get_bits_16(decoded_data, 18, 16);
+
+    string_cat_printf(result, "Facility: %u\r\n", fc);
+    string_cat_printf(result, "Card: %lu\r\n", card_id);
+    string_cat_printf(result, "Data: ");
+    for(size_t i = 0; i < PARADOX_DECODED_DATA_SIZE; i++) {
+        string_cat_printf(result, "%02X", decoded_data[i]);
+    }
+};
+
+void protocol_paradox_render_brief_data(ProtocolParadox* protocol, string_t result) {
+    uint8_t* decoded_data = protocol->data;
+
+    uint8_t fc = bit_lib_get_bits(decoded_data, 10, 8);
+    uint16_t card_id = bit_lib_get_bits_16(decoded_data, 18, 16);
+
+    string_cat_printf(result, "ID: %03u,%05u", fc, card_id);
+};
+
+bool protocol_paradox_write_data(ProtocolParadox* protocol, void* data) {
+    LFRFIDWriteRequest* request = (LFRFIDWriteRequest*)data;
+    bool result = false;
+
+    protocol_paradox_encode(protocol->data, (uint8_t*)protocol->encoded_data);
+
+    if(request->write_type == LFRFIDWriteTypeT5577) {
+        request->t5577.block[0] = LFRFID_T5577_MODULATION_FSK2a | LFRFID_T5577_BITRATE_RF_50 |
+                                  (3 << LFRFID_T5577_MAXBLOCK_SHIFT);
+        request->t5577.block[1] = bit_lib_get_bits_32(protocol->encoded_data, 0, 32);
+        request->t5577.block[2] = bit_lib_get_bits_32(protocol->encoded_data, 32, 32);
+        request->t5577.block[3] = bit_lib_get_bits_32(protocol->encoded_data, 64, 32);
+        request->t5577.blocks_to_write = 4;
+        result = true;
+    }
+    return result;
```

Comment:
straight `return true/false`?

Reply:
A single exit point from a function is highly desirable.